False error oas3-valid-schema-example  reference resolves to more than one schema

> For support questions, please use the [Stoplight Discord Community](https://discord.com/invite/stoplight). This repository's issues are reserved for feature requests and bug reports. If you are unsure if you are experiencing a bug, our Discord is a great place to start.
>
> **Please delete this section, any any sections below that you don't use, before creating the issue.**

**Describe the bug**
Spectral is generating a strange error when a schema has a property with the name `id`.

**To Reproduce**

1. Given this OpenAPI/AsyncAPI document 

```yaml
openapi: 3.0.3
info:
  title: Spectral Issue
  description: >-
    error oas3-valid-schema-example reference "bf23bc970b78d27691e8" resolves to more than one schema
  version: 0.1.0
  contact:
    name: Apiture

servers:
  - url: /spectral/issues

tags:
  - name: Spectral Issues
    description: Spectral Issues

paths:

  /path:
    get:
      operationId: getResource
      description: Get a resource
      tags:
        - Spectral Issues
      responses:
        '200':
          description: OK.
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/collection"


components:

  schemas:

    item:
      type: object
      properties:
        id:
          description: >-
            The unique identifier for this account resource.
            This is an immutable opaque string.
          readOnly: true
          type: string
          pattern: '^[-_:.~$a-zA-Z0-9]+$'
          minLength: 6
          maxLength: 48
          example: bf23bc970b78d27691e8
        url:
          description: >-
            The URL of this account instance. See the [`getAccount`](#op-getAccount) operation.
          readOnly: true
          maxLength: 256
          type: string
          format: uri
          example: https://api.apiture.com/banking/accounts/bf23bc970b78d27691e8
      example:
        id: bf23bc970b78d27691e8
        url: https://api.apiture.com/banking/accounts/bf23bc970b78d27691e8

    collection:
      required:
        - items
      type: object
      allOf:
        - type: object
          properties:
            items:
              description: The array of items in this page of accounts.
              readOnly: true
              type: array
              items:
                $ref: '#/components/schemas/item'
      example:
        items:
          - id: bf23bc970b78d27691e8
            url: https://api.example.com/spectral/issues/bf23bc970b78d27691e8
          - id: 8d27691e8bf23bc970b7
            url: https://api.example.com/spectral/issues/8d27691e8bf23bc970b7
```

and this `.spectral.yaml`
3. Run this CLI command: `spectral lint -r ./.spectral.yaml openapi.yaml`
4. See error

```
spectral lint -r ./.spectral.yaml openapi.yaml

./spectral-issues/openapi.yaml
 76:15  error  oas3-valid-schema-example  reference "bf23bc970b78d27691e8" resolves to more than one schema  components.schemas.collection.example

✖ 1 problem (1 error, 0 warnings, 0 infos, 0 hints)

```

**Expected behavior**

There should be no error here.

**Environment (remove any that are not applicable):**
 - Library version: spectral-cli 6.0.0 (newer versions are not working for me; see #2080 )
 - OS: Mac OS
 - Browser: N/A

--- a/packages/rulesets/src/oas/functions/__tests__/oasExample.test.ts
+++ b/packages/rulesets/src/oas/functions/__tests__/oasExample.test.ts
@@ -114,6 +114,9 @@ describe('oasExample', () => {
                     type: 'object',
                     required: ['id', 'url'],
                     properties: {
+                      'x-stoplight': {
+                        id: 'abc',
+                      },
                       id: {
                         type: 'string',
                       },
@@ -121,6 +124,9 @@ describe('oasExample', () => {
                         type: 'string',
                         format: 'uri',
                       },
+                    },
+                    'x-stoplight': {
+                      id: 'abc',
                     },
                     example: {
                       id: 'bf23bc970b78d27691e8',

--- a/packages/rulesets/src/oas/functions/oasExample.ts
+++ b/packages/rulesets/src/oas/functions/oasExample.ts
@@ -111,17 +111,37 @@ function* getSchemaValidationItems(
   }
 }
 
+const KNOWN_TRAVERSE_KEYWORDS = [
+  /* eslint-disable @typescript-eslint/no-unsafe-argument */
+  ...Object.keys(traverse['keywords']),
+  ...Object.keys(traverse['arrayKeywords']),
+  ...Object.keys(traverse['propsKeywords']),
+  /* eslint-enable @typescript-eslint/no-unsafe-argument */
+];
+
 /**
- * Modifies 'schema' (and all its sub-schemas) to remove all "example" fields.
+ * Modifies 'schema' (and all its sub-schemas) to remove all id fields from non-schema objects
  * In this context, "sub-schemas" refers to all schemas reachable from 'schema'
  * (e.g. properties, additionalProperties, allOf/anyOf/oneOf, not, items, etc.)
- * @param schema the schema to be "de-examplified"
- * @returns 'schema' with example fields removed
+ * @param schema the schema to be sanitized
+ * @returns 'schema' with id fields removed
  */
-function deExamplify(schema: Record<string, unknown>): void {
-  traverse(schema, <traverse.Callback>(fragment => {
-    if ('example' in fragment) {
-      delete fragment.example;
+function cleanSchema(schema: Record<string, unknown>): void {
+  traverse(schema, { allKeys: true }, <traverse.Callback>((
+    fragment,
+    jsonPtr,
+    rootSchema,
+    parentJsonPtr,
+    parentKeyword,
+  ) => {
+    if (parentKeyword === void 0 || KNOWN_TRAVERSE_KEYWORDS.includes(parentKeyword)) return;
+
+    if ('id' in fragment) {
+      delete fragment.id;
+    }
+
+    if ('$id' in fragment) {
+      delete fragment.id;
     }
   }));
 }
@@ -165,11 +185,11 @@ export default createRulesetFunction<Record<string, unknown>, Options>(
       delete schemaOpts.schema.required;
     }
 
-    // Make a deep copy of the schema and then remove all the "example" fields from it.
+    // Make a deep copy of the schema and then remove all objects containing id or $id and that are not schema objects.
     // This is to avoid problems down in "ajv" which does the actual schema validation.
     // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
     schemaOpts.schema = JSON.parse(JSON.stringify(schemaOpts.schema));
-    deExamplify(schemaOpts.schema);
+    cleanSchema(schemaOpts.schema);
 
     for (const validationItem of validationItems) {
       const result = oasSchema(validationItem.value, schemaOpts, {

